### PR TITLE
do not rely on udevadm cli

### PR DESCRIPTION
**Pull request by leseb**
Related #2989

### CODE REVIEW

**Comment by mulkieran** [comment]
@leseb Thanks for the PR! We have two targets in our Makefile, "fmt" and "clippy" and we enforce their rules in our CI. Running those in your development environment will save you the trouble of unnecessary round trips w/ our CI.

File: src/engine/strat_engine/udev.rs [hunk 1]
```diff
@@ -157,10 +157,60 @@ where
...
+
+    // If the file exists then the queue is not empty and we return an error
+    if udev_queue_file_path.exists() {
+        Err(StratisError::Msg(format!(
+            "Udev queue file {} still exists",
+            UDEV_QUEUE_FILE_PATH
+        )))
+    } else {
+        Ok(true)
+    }
```
Comment (jbaublitz) [inline src/engine/strat_engine/udev.rs:184]: Mainly stylistic, but I'd think this return type should either be `StratisResult<()>` or `bool`. I don't see any case where `Ok(false)` is returned.

**Review by jbaublitz — COMMENTED** [review]
I'm unsure whether this conflicts with the udev sync work, but if that PR is still using udevadm and CI doesn't fail for this PR, this may be okay to merge prior to that.

File: src/engine/strat_engine/cmd.rs [hunk 2]
```diff
@@ -14,6 +14,9 @@
 // the existence of the file is checked before the command is invoked, and
 // an explicit error is returned if the executable can not be found.
 
+use libc::c_uint;
+use libcryptsetup_rs::SafeMemHandle;
+use serde_json::Value;
```
Comment (jbaublitz) [inline src/engine/strat_engine/cmd.rs:19]: We generally want the imports to be separated into the following categories:

`std`

external crates (like libc, serde_json)

crates that are external that we develop (like devicemapper)

`crate`

all separated by a newline so they're ordered appropriately within those categories.

File: src/engine/strat_engine/udev.rs [hunk 3]
```diff
@@ -157,10 +157,60 @@ where
...
         .scan_devices()?
         .filter(|dev| dev.is_initialized())
         .find(|x| x.devnode().map_or(false, |d| **device_path == *d))
         .map(|ref d| f(&UdevEngineDevice::from(d))))
 }
+
+/// Return true if udev queue is empty and no events are being processed.
+#[cfg(test)]
+pub fn is_udev_queue_empty() -> StratisResult<()> {
+    // Path of udev queue's file, typically this file is empty, it just asses that udev events are still
```
Comment (jbaublitz) [inline src/engine/strat_engine/udev.rs:170]: This should be "it just assesses"

**Review by jbaublitz — CHANGES_REQUESTED** [review]
After these are handled, I think we can accept this PR.

**Comment by jbaublitz** [comment]
Oops, missed that CI is failing. We'll have to investigate that first too @leseb.

**Comment by jbaublitz** [comment]
@leseb Looks like we've fixed all of the failures by doing that except for `check_with_tang` which appears to not have been modified. If you apply the same changes there, I have a feeling that we should be good to go with CI.

**Comment by mulkieran** [comment]
I think we're trying to fix up some CI problems, see #3211

**Comment by mulkieran** [comment]
The CI problems seem to be caused by the non-cli implementation of settle starting and finishing too fast. We aren't sure, but it might be that the udev synchronization work would resolve this problem.

**Comment by leseb** [comment]
> The CI problems seem to be caused by the non-cli implementation of settle starting and finishing too fast. We aren't sure, but it might be that the udev synchronization work would resolve this problem.

Agreed, hence the retry in this PR. This might be environment specific since the runners are small VMs, something we might not encounter on larger machines. Also, AFAIC this is only executed during the unit tests so retry is not uncommon in CI :).

**Comment by leseb** [comment]
Closing this PR since it will soon be superseded by another implementation (udev synchronization improvements). We may re-open in the future if necessary. Thanks @mulkieran @jbaublitz for your support on this.

**Pull request closed without merging**